### PR TITLE
Amoc 2.0.0 rework

### PR DESCRIPTION
**Pull request by DenysGonchar**

### CODE REVIEW

File: src/amoc_config/amoc_config.erl [hunk 1]
```diff
@@ -37,127 +20,17 @@
...
-parse_scenario_settings(Module) when is_atom(Module) ->
-    try get_scenario_configuration(Module) of
-        Config -> parse_scenario_settings(Config)
-    catch
-        _:_ -> {error, invalid_module}
-    end;
-parse_scenario_settings(Config) ->
-    ParametersVerification = [get_value_and_verify(C) || C <- Config],
-    case [{N, V, R} || {N, V, {false, R}} <- ParametersVerification] of
+    case ets:lookup(amoc_config,Name) of
```
Comment (michalwski) [inline src/amoc_config/amoc_config.erl]:
```suggestion
    case ets:lookup(amoc_config, Name) of
```

File: src/amoc_config/amoc_config.erl [hunk 2]
```diff
@@ -37,127 +20,17 @@
...
-    {Name, DefaultValue, VerificationMethod}.
-
-verification_method(Module, Atom) ->
-    code:ensure_loaded(Module),
-    case erlang:function_exported(Module, Atom, 1) of
-        true -> fun Module:Atom/1;
-        false -> Atom
+            ?LOG_ERROR("no scenario setting ~p", [Name]),
+            Default;
+        [{Name,undefined}]->
```
Comment (michalwski) [inline src/amoc_config/amoc_config.erl]:
```suggestion
        [{Name, undefined}]->
```

File: src/amoc_config/amoc_config.erl [hunk 3]
```diff
@@ -37,127 +20,17 @@
...
+            Default;
+        [{Name,Value}] ->
```
Comment (michalwski) [inline src/amoc_config/amoc_config.erl]:
```suggestion
        [{Name, Value}] ->
```

File: src/amoc_config/amoc_config.erl [hunk 4]
```diff
@@ -37,127 +20,17 @@
...
+            Value;
+        InvalidLookupRet->
+            ?LOG_ERROR("invalid lookup return value ~p ~p", [Name,InvalidLookupRet]),
```
Comment (michalwski) [inline src/amoc_config/amoc_config.erl]:
```suggestion
            ?LOG_ERROR("invalid lookup return value ~p ~p", [Name, InvalidLookupRet]),
```

File: src/amoc_config/amoc_config_env.erl [hunk 5]
```diff
@@ -0,0 +1,55 @@
+%%==============================================================================
+%% Copyright 2016 Erlang Solutions Ltd.
```
Comment (michalwski) [inline src/amoc_config/amoc_config_env.erl]:
```suggestion
%% Copyright 2019 Erlang Solutions Ltd.
```

File: src/amoc_config/amoc_config_scenario.erl [hunk 6]
```diff
@@ -0,0 +1,202 @@
+%%==============================================================================
+%% Copyright 2016 Erlang Solutions Ltd.
```
Comment (michalwski) [inline src/amoc_config/amoc_config_scenario.erl]:
```suggestion
%% Copyright 2019 Erlang Solutions Ltd.
```

File: src/amoc_controller.erl [hunk 7]
```diff
@@ -305,89 +246,21 @@ start_scenario(Scenario, UserIds, State) ->
...
-        true ->
-            case Scenario:init() of
-                ok -> {ok, ok};
-                RetValue -> RetValue
+-spec init_scenario(amoc:scenario(), amoc:do_opts()) -> any().
+init_scenario(Scenario, Opts) ->
+    ScenarioConfig = proplists:get_value(config, Opts, []),
+    case amoc_config_scenario:parse_scenario_settings(Scenario, ScenarioConfig) of
+        ok ->
+            case erlang:function_exported(Scenario, init, 0) of
```
Comment (michalwski) [inline src/amoc_controller.erl]: I'd like to see this case extracted to a separate function named `call_scenario_init` or sth like that. The reason why I want this is that currently have 3 levels of cases in single function.

File: src/amoc_scenario.erl [hunk 8]
```diff
@@ -12,12 +27,128 @@
...
+                             protected,
+                             {read_concurrency, true}]).
+
+-spec find_scenario_modules() -> [module()].
+find_scenario_modules() ->
+    AllBeamFiles = [File || Path <- code:get_path(), File <- filelib:wildcard("*.beam", Path)],
+    AllModules = [list_to_atom(filename:rootname(BeamFile)) || BeamFile <- AllBeamFiles],
+    ok = code:ensure_modules_loaded(AllModules),
+    AllScenarios = [Module || Module <- erlang:loaded(), is_scenario(Module)],
+    [ets:insert(amoc_scenarios, {Scenario, scenario, preloaded}) || Scenario <- AllScenarios].
```
Comment (michalwski) [inline src/amoc_scenario.erl]: Why do we need to have all possible scenarios preloaded?

File: test/amoc_api_scenarios_handler_SUITE.erl [hunk 9]
```diff
@@ -80,18 +76,15 @@ post_scenarios_returns_200_and_compile_error_when_scenario_source_not_valid(_Con
...
+              "\",\n  [{1,erl_lint,undefined_module}]}]">>,
+    ?assertEqual({[{<<"compile">>,Error}]},Body).
 
-post_scenarios_returns_200_and_when_scenario_valid(Config) ->
+post_scenarios_returns_200_and_when_scenario_valid(_Config) ->
     %% given
-    ScenarioFile = data(Config, ?SAMPLE_SCENARIO_S),
-    {ok, ScenarioContent} = file:read_file(ScenarioFile),
+    ScenarioContent = ?DUMMY_SCENARIO_MODULE(?SAMPLE_SCENARIO_A),
+    io:format("~n!!!!!!!!!!!!!!!!!!! ~p ~n",[ScenarioContent]),
```
Comment (michalwski) [inline test/amoc_api_scenarios_handler_SUITE.erl]: Is this a leftover?

File: test/amoc_config_scenario_SUITE.erl [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+    ?assertEqual(def1, amoc_config:get(var1, def1)),
+    ?assertEqual(def2, amoc_config:get(var2)),
+    ?assertEqual(def2, amoc_config:get(var2, some_atom)),
+    ?assertEqual(val5, amoc_config:get(var5)),
+    ?assertEqual(val5, amoc_config:get(var5, some_atom)).
+
+get_module_attributes(_) ->
+    Result = amoc_config_scenario:get_module_attributes(?MODULE),
+    ExpectedResult = [{var1,"var1"},
+                      {var2,"var2",def2},
```
Comment (michalwski) [inline test/amoc_config_scenario_SUITE.erl]:
```suggestion
                      {var2, "var2", def2},
```

File: test/amoc_config_scenario_SUITE.erl [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+    ?assertEqual(undefined, amoc_config:get(var1)),
+    ?assertEqual(def1, amoc_config:get(var1, def1)),
+    ?assertEqual(def2, amoc_config:get(var2)),
+    ?assertEqual(def2, amoc_config:get(var2, some_atom)),
+    ?assertEqual(val5, amoc_config:get(var5)),
+    ?assertEqual(val5, amoc_config:get(var5, some_atom)).
+
+get_module_attributes(_) ->
+    Result = amoc_config_scenario:get_module_attributes(?MODULE),
+    ExpectedResult = [{var1,"var1"},
```
Comment (michalwski) [inline test/amoc_config_scenario_SUITE.erl]:
```suggestion
    ExpectedResult = [{var1, "var1"},
```

File: test/amoc_config_scenario_SUITE.erl [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+                      {var2,"var2",def2},
+                      {var3,"var3",3,positive_integer},
```
Comment (michalwski) [inline test/amoc_config_scenario_SUITE.erl]:
```suggestion
                      {var3, "var3", 3, positive_integer},
```

File: test/amoc_config_scenario_SUITE.erl [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+                      {var4,"var4",def4,[def4,another_atom]},
```
Comment (michalwski) [inline test/amoc_config_scenario_SUITE.erl]:
```suggestion
                      {var4, "var4", def4, [def4, another_atom]},
```

File: test/amoc_config_scenario_SUITE.erl [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+                      {var5,"var5",def5,test_verification_function},
```
Comment (michalwski) [inline test/amoc_config_scenario_SUITE.erl]:
```suggestion
                      {var5, "var5", def5, test_verification_function},
```

File: test/amoc_config_scenario_SUITE.erl [hunk 15]
```diff
@@ -0,0 +1,162 @@
...
+                      {var6,"var6",def6, fun ?MODULE:test_verification_function/1}],
```
Comment (michalwski) [inline test/amoc_config_scenario_SUITE.erl]:
```suggestion
                      {var6, "var6", def6, fun ?MODULE:test_verification_function/1}],
```

File: test/amoc_config_scenario_SUITE.erl [hunk 16]
```diff
@@ -0,0 +1,162 @@
...
+    ScenarioAttributes = correct_scenario_attributes(),
+    given_scenario_parameters_not_set(ScenarioAttributes),
+    Result = process_scenario_attributes(ScenarioAttributes,[]),
+    assert_settings(Result, ScenarioAttributes).
+
+process_scenario_config_shadows_default_values(_) ->
+    ScenarioAttributes = correct_scenario_attributes(),
+    given_scenario_parameters_set(ScenarioAttributes),
+    AnotherScenarioAttributes = another_correct_scenario_attributes(),
+    AdditionalAttribute = {additional_param,"", value, none},
```
Comment (michalwski) [inline test/amoc_config_scenario_SUITE.erl]:
```suggestion
    AdditionalAttribute = {additional_param, "", value, none},
```

File: test/amoc_config_scenario_SUITE.erl [hunk 17]
```diff
@@ -0,0 +1,162 @@
...
+    set_env(additional_param, another_value),
+    Result = process_scenario_attributes(
+        [AdditionalAttribute | AnotherScenarioAttributes],
+        [{additional_param, yet_another_value}]),
+    assert_settings(Result, [{additional_param,"", yet_another_value, none} | ScenarioAttributes]).
```
Comment (michalwski) [inline test/amoc_config_scenario_SUITE.erl]:
```suggestion
    assert_settings(Result, [{additional_param, "", yet_another_value, none} | ScenarioAttributes]).
```

File: src/amoc_scenario.erl (hunk 8 above)
Comment (DenysGonchar) [reply src/amoc_scenario.erl]: it's required for dynamic addition of the new amoc node during the scenario execution

File: test/amoc_api_scenarios_handler_SUITE.erl (hunk 9 above)
Comment (DenysGonchar) [reply test/amoc_api_scenarios_handler_SUITE.erl]: yes.

**Comment by DenysGonchar** [comment]
> That's a huge change. Splitting the `amoc_config` into 2 modules, one for scenario variables the other for generic settings makes sense. After reading through the code I think it'd be clearer and more obvious to the reader if currently named `amoc_config` module was called `amoc_scenario_config` and the new `amoc_config_env` was just `amoc_config`. What do you think?

`amoc_config` module has the only 2 interfaces that users must use to extract scenario parameters, and users are not supposed to use `amoc_config_env` directly. so to avoid any collisions i would prefer to keep  `amoc_config_env` name.

File: src/amoc_controller.erl (hunk 7 above)
Comment (DenysGonchar) [reply src/amoc_controller.erl]: amoc_controller will be heavily reworked in the next PR. and this code won't be there any more

**Pull request merged**